burn-core/src/nn/lstm/lstm.rs

```diff
@@ -0,0 +1,322 @@
+            d_output,
+            self.bias,
+            self.initializer.clone(),
+        );
+        let output_gate = gate_controller::GateController::new(
+            self.d_input,
+            d_output,
+            self.bias,
+            self.initializer.clone(),
+        );
+        let cell_gate = gate_controller::GateController::new(
+            self.d_input,
+            d_output,
+            self.bias,
+            self.initializer.clone(),
+        );
+
+        Lstm {
+            input_gate,
+            forget_gate,
+            output_gate,
+            cell_gate,
+            batch_size: self.batch_size,
+            d_hidden: self.d_hidden,
+        }
+    }
+
+    /// Initialize a new [lstm](lstm) module with a [record](LstmRecord).
+    pub fn init_with<B: Backend>(&self, record: LstmRecord<B>) -> Lstm<B> {
+        let linear_config = LinearConfig {
+            d_input: self.d_input,
+            d_output: self.d_hidden,
+            bias: self.bias,
+            initializer: self.initializer.clone(),
+        };
+
+        Lstm {
+            input_gate: gate_controller::GateController::new_with(
+                &linear_config,
+                record.input_gate,
+            ),
+            forget_gate: gate_controller::GateController::new_with(
+                &linear_config,
+                record.forget_gate,
+            ),
+            output_gate: gate_controller::GateController::new_with(
+                &linear_config,
+                record.output_gate,
+            ),
+            cell_gate: gate_controller::GateController::new_with(&linear_config, record.cell_gate),
+            batch_size: self.batch_size,
+            d_hidden: self.d_hidden,
+        }
+    }
+}
+
+impl<B: Backend> Lstm<B> {
+    /// Applies the forward pass on the input tensor. This LSTM implementation
+    /// returns the cell state and hidden state for each element in a sequence (i.e., across `seq_length`),
+    /// producing 3-dimensional tensors where the dimensions represent [batch_size, seq_length, hidden_size].
+    ///
+    /// Parameters:
+    ///     batched_input: The input tensor of shape [batch_size, seq_length, input_size].
+    ///     state: An optional tuple of tensors representing the initial cell state and hidden state.
+    ///            Each state tensor has shape [batch_size, hidden_size].
+    ///            If no initial state is provided, these tensors are initialized to zeros.
+    ///
+    /// Returns:
+    ///     A tuple of tensors, where the first tensor represents the cell states and
+    ///     the second tensor represents the hidden states for each sequence element.
+    ///     Both output tensors have the shape [batch_size, seq_length, hidden_size].
+    pub fn forward(
+        &mut self,
+        batched_input: Tensor<B, 3>,
+        state: Option<(Tensor<B, 2>, Tensor<B, 2>)>,
+    ) -> (Tensor<B, 3>, Tensor<B, 3>) {
+        let seq_length = batched_input.shape().dims.to_vec()[1];
```

Comment:
Just noticed the dims  (array) are converted to vec(). Probably more efficient without `to_vec()`.